# pandas

# 3. 데이터 입출력 

## 3.1 파일로 출력

- 데이터를 csv 파일로 출력할 땐 to_csv()메서드 활용 
- 첫 인자로는 파일 경로를 입력 
    - 아래에서 만들 DataFrame의 index는 의미 없는 값이므로 출력할 때 고려하지 않음 
    - to_csv()의 기본값 인자는 index의 default=True이니 index=False로 설정해야함 

In [2]:
import pandas as pd

data = {
    'c1': [1,2,'누락'],
    'c2': [1.11, "", 3.33],
    'c3': ['one', 'two', 'three'],
}

df = pd.DataFrame(data)
df

c1    c2     c3
0   1  1.11    one
1   2          two
2  누락  3.33  three

In [4]:
df.to_csv('sample.csv', index=False)

In [5]:
df.to_csv('sample2.csv', index=True)

In [6]:
df.to_csv('sample3.csv', index=True, header=False)

## 3.2 csv파일에서 데이터 불러오기 
- pandas.read_csv()메서드 활용 
- 첫 번째 인수로 '파일 이름.확장자' 문자열 삽입

1. 1번 파일 불러오기

In [7]:
df = pd.read_csv('sample.csv')
df

c1    c2     c3
0   1  1.11    one
1   2   NaN    two
2  누락  3.33  three

2. index=True 로 저장한 파일 불러와보기
- index=True에서 index가 첫 번재 열로 csv 파일에 저장되어버림 -> 다시 불러올 때 unnamed 칼럼명이 붙으면서 첫 번째 열로 읽혀짐

In [8]:
df = pd.read_csv('sample2.csv')
df

Unnamed: 0  c1    c2     c3
0           0   1  1.11    one
1           1   2   NaN    two
2           2  누락  3.33  three

In [11]:
df2 = pd.read_csv('sample2.csv', index_col=0)
df2

c1    c2     c3
0   1  1.11    one
1   2   NaN    two
2  누락  3.33  three

3. column 지정하지 않은 sample3.csv 불러와보기 

In [12]:
df = pd.read_csv('sample3.csv')
df

0   1  1.11    one
0  1   2   NaN    two
1  2  누락  3.33  three

In [13]:
df = pd.read_csv('sample3.csv', names=['c1', 'c2', 'c3'])
df

c1    c2     c3
0   1  1.11    one
1   2   NaN    two
2  누락  3.33  three

## 3.3 pokemon.csv 파일 읽어오기 

In [18]:
df = pd.read_csv('pokemon.csv')
df

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

- 행 건너뛰기: 파일 중 건너 뛰어야 할 상단 행이 있으면 skiprows 인수 사용
- 리스트나 range 통해 인수 전달 

- 데이터로 불러올 자료 안 특정한 값을 NaN으로 취급하고 싶으면 
- na_values 인수에 NaN 값으로 취급할 값을 넣어줌

- na_rep 키워드 인수 사용해서 NaN 표시값 바꾸기 
- NaN 값을 '누락'으로 변경해서 저장해보기

- 경로가 아닌 url을 인수로 전달하면 해당 파일을 판다스가 직접 다운로드 해옴 

In [29]:
titanic = pd.read_csv("https://storage.googleapis.com/tf-datasets/titanic/train.csv") # 따로 저장되지는 않음
titanic

survived     sex   age  n_siblings_spouses  parch     fare   class  \
0           0    male  22.0                   1      0   7.2500   Third   
1           1  female  38.0                   1      0  71.2833   First   
2           1  female  26.0                   0      0   7.9250   Third   
3           1  female  35.0                   1      0  53.1000   First   
4           0    male  28.0                   0      0   8.4583   Third   
..        ...     ...   ...                 ...    ...      ...     ...   
622         0    male  28.0                   0      0  10.5000  Second   
623         0    male  25.0                   0      0   7.0500   Third   
624         1  female  19.0                   0      0  30.0000   First   
625         0  female  28.0                   1      2  23.4500   Third   
626         0    male  32.0                   0      0   7.7500   Third   

        deck  embark_town alone  
0    unknown  Southampton     n  
1          C    Cherbourg     n  
2    unknown  Southampton     y  
3          C  Southampton     n  
4    unknown   Queenstown     y  
..       ...          ...   ...  
622  unknown  Southampton     y  
623  unknown  Southampton     y  
624        B  Southampton     y  
625  unknown  Southampton     n  
626  unknown   Queenstown     y  

[627 rows x 10 columns]

## 3.4 head() / tail()

- 앞이나 뒤의 특정 개수만 보고싶을 경우 사용 
- 메서드 인수로 출력할 행의 개수를 입력하면 됨 

In [19]:
# 상위행 5개(default)
df.head(20)

#                       Name Type 1  Type 2  Total  HP  Attack  Defense  \
0    1                  Bulbasaur  Grass  Poison    318  45      49       49   
1    2                    Ivysaur  Grass  Poison    405  60      62       63   
2    3                   Venusaur  Grass  Poison    525  80      82       83   
3    3      VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4    4                 Charmander   Fire     NaN    309  39      52       43   
5    5                 Charmeleon   Fire     NaN    405  58      64       58   
6    6                  Charizard   Fire  Flying    534  78      84       78   
7    6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   
8    6  CharizardMega Charizard Y   Fire  Flying    634  78     104       78   
9    7                   Squirtle  Water     NaN    314  44      48       65   
10   8                  Wartortle  Water     NaN    405  59      63       80   
11   9                  Blastoise  Water     NaN    530  79      83      100   
12   9    BlastoiseMega Blastoise  Water     NaN    630  79     103      120   
13  10                   Caterpie    Bug     NaN    195  45      30       35   
14  11                    Metapod    Bug     NaN    205  50      20       55   
15  12                 Butterfree    Bug  Flying    395  60      45       50   
16  13                     Weedle    Bug  Poison    195  40      35       30   
17  14                     Kakuna    Bug  Poison    205  45      25       50   
18  15                   Beedrill    Bug  Poison    395  65      90       40   
19  15      BeedrillMega Beedrill    Bug  Poison    495  65     150       40   

    Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0        65       65     45           1      False  
1        80       80     60           1      False  
2       100      100     80           1      False  
3       122      120     80           1      False  
4        60       50     65           1      False  
5        80       65     80           1      False  
6       109       85    100           1      False  
7       130       85    100           1      False  
8       159      115    100           1      False  
9        50       64     43           1      False  
10       65       80     58           1      False  
11       85      105     78           1      False  
12      135      115     78           1      False  
13       20       20     45           1      False  
14       25       25     30           1      False  
15       90       80     70           1      False  
16       20       20     50           1      False  
17       25       25     35           1      False  
18       45       80     75           1      False  
19       15       80    145           1      False

In [20]:
# 하위행 5개(default)
df.tail()


#                 Name   Type 1 Type 2  Total  HP  Attack  Defense  \
795  719              Diancie     Rock  Fairy    600  50     100      150   
796  719  DiancieMega Diancie     Rock  Fairy    700  50     160      110   
797  720  HoopaHoopa Confined  Psychic  Ghost    600  80     110       60   
798  720   HoopaHoopa Unbound  Psychic   Dark    680  80     160       60   
799  721            Volcanion     Fire  Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True

## 3.5 nunique()
- *고유한 값의 개수를 계산할 때 사용*
- DF객체는 nunique()메서드에 대해 각 칼럼마다 갖는 고유할 값을 Series로 반환

In [21]:
df.nunique()

#             721
Name          800
Type 1         18
Type 2         18
Total         200
HP             94
Attack        111
Defense       103
Sp. Atk       105
Sp. Def        92
Speed         108
Generation      6
Legendary       2
dtype: int64

## 3.6 count()

- 칼럼마다의 데이터 개수 계산 = 행개수를 센다 \
행 = rows = 데이터
- *NaN인 값에 대해서는 개수 포함 안 함*
- 칼럼마다 개수를 보여줘야 해서 Series로 반환 
- DF 객체를 인수로 해서 len()함수를 사용하면 -> row index의 전체 크기를 알려줌 (즉, 행 개수 반환)
    - nan값과 관계없이 전체를 세기 때문에 단 하나의 정수 값을 반환 

In [23]:
print(df.count())

#             800
Name          800
Type 1        800
Type 2        414
Total         800
HP            800
Attack        800
Defense       800
Sp. Atk       800
Sp. Def       800
Speed         800
Generation    800
Legendary     800
dtype: int64


In [25]:
len(df)


800

## 3.7 value_counts()

- *고유값 개수 카운팅*
- 다양한 옵션 설정 가능
    - ascending=True : 오름차순 정렬
    - dropna = True : Na값을 집계에 포함시킴 
    - normalize=True : 비율 계산

In [24]:
df['Type 1'].value_counts()

Type 1
Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ground       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: count, dtype: int64

In [27]:
df['Type 1'].value_counts(normalize=True).round(2) # normalize - 고유값 등장 비율 계산, 각 값이 차지하는 비율

Type 1
Water       0.14
Normal      0.12
Grass       0.09
Bug         0.09
Psychic     0.07
Fire        0.06
Electric    0.06
Rock        0.06
Dragon      0.04
Ground      0.04
Ghost       0.04
Dark        0.04
Poison      0.04
Steel       0.03
Fighting    0.03
Ice         0.03
Fairy       0.02
Flying      0.00
Name: proportion, dtype: float64

In [28]:
df[['Attack', 'HP']].value_counts() # 서로 겹치는 값에 대한 데이터를 반환

Attack  HP 
100     100    9
65      50     9
50      50     8
60      60     8
85      60     7
              ..
68      86     1
        55     1
67      60     1
        45     1
190     106    1
Name: count, Length: 517, dtype: int64

## 3.8 dtypes

- titanic의 각 컬럼에 대해 dtype을 조회 가능 
- object type은 주로 문자열 혹은 문자열+숫자의 혼합일 때 주로 나타남


In [31]:
titanic.dtypes # 각 열에 대해 조회

survived                int64
sex                    object
age                   float64
n_siblings_spouses      int64
parch                   int64
fare                  float64
class                  object
deck                   object
embark_town            object
alone                  object
dtype: object

## 3.9 describe()

- 수치 값을 갖는 DataFrame의 각 컬럼에 대해 *count, mean, std, min, 25%, median(50%), 75%, max*에 대한 모든 통계를 구해줌


In [32]:
titanic.describe()

survived         age  n_siblings_spouses       parch        fare
count  627.000000  627.000000          627.000000  627.000000  627.000000
mean     0.387560   29.631308            0.545455    0.379585   34.385399
std      0.487582   12.511818            1.151090    0.792999   54.597730
min      0.000000    0.750000            0.000000    0.000000    0.000000
25%      0.000000   23.000000            0.000000    0.000000    7.895800
50%      0.000000   28.000000            0.000000    0.000000   15.045800
75%      1.000000   35.000000            1.000000    0.000000   31.387500
max      1.000000   80.000000            8.000000    5.000000  512.329200

# 4. 결측치(None) 다루기 

- series때 다뤘던 것과 마찬가지로 동일한 메소드를 사용할 수 있음
- 특히 .isnull().sum()과 같은 결측치 확인 명령어는 알아두면 유용


In [33]:
import numpy as np

# 결측치가 포함된 샘플 데이터셋 생성
data = {
    "Name": ["Alice", "Bob", "Charlie", "David", "Eva"],
    "Age": [25, np.nan, 35, 40, np.nan],
    "Salary": [50000, 60000, np.nan, 80000, 90000],
    "Department": ["HR", "Finance", np.nan, "IT", "Marketing"]
}

df = pd.DataFrame(data)

## 4.1 isnull() / notnull()

In [34]:
df.isnull()

Name    Age  Salary  Department
0  False  False   False       False
1  False   True   False       False
2  False  False    True        True
3  False  False   False       False
4  False   True   False       False

In [35]:
df.isnull().sum() # 각열의 True 값을 합해줌 - 각 열의 null 의 개수를 알수 있음

Name          0
Age           2
Salary        1
Department    1
dtype: int64

In [36]:
df.notnull()

Name    Age  Salary  Department
0  True   True    True        True
1  True  False    True        True
2  True   True   False       False
3  True   True    True        True
4  True  False    True        True

In [37]:
df.notnull().sum() # 각열의 값이 존재하는 것의 개수

Name          5
Age           3
Salary        4
Department    4
dtype: int64

## 4.2 dropna() / fillna()

- dropna() : 결측치가 존재하는 행을 제외 - none 값이 존재하는 행을 지우기
- fillna('값') : 결측치를 원하는 값으로 채울 수 있음 

In [39]:
df

Name   Age   Salary Department
0    Alice  25.0  50000.0         HR
1      Bob   NaN  60000.0    Finance
2  Charlie  35.0      NaN        NaN
3    David  40.0  80000.0         IT
4      Eva   NaN  90000.0  Marketing

In [40]:
df2 = df.dropna() # 원본에 영향을 주지않음
df2

Name   Age   Salary Department
0  Alice  25.0  50000.0         HR
3  David  40.0  80000.0         IT

In [41]:
df.fillna('누락') # 각 데이터의 특성, 타입을 고려하여 채워줘야 하기때문에 한번에 때려넣는 식은 잘 사용하지 않음

Name   Age   Salary Department
0    Alice  25.0  50000.0         HR
1      Bob    누락  60000.0    Finance
2  Charlie  35.0       누락         누락
3    David  40.0  80000.0         IT
4      Eva    누락  90000.0  Marketing

In [42]:
df

Name   Age   Salary Department
0    Alice  25.0  50000.0         HR
1      Bob   NaN  60000.0    Finance
2  Charlie  35.0      NaN        NaN
3    David  40.0  80000.0         IT
4      Eva   NaN  90000.0  Marketing

In [43]:
df['Age'].fillna('누락1') # 시리즈로 반환

0    25.0
1     누락1
2    35.0
3    40.0
4     누락1
Name: Age, dtype: object

In [44]:
df['Age'] = df['Age'].fillna('누락1')
df

Name   Age   Salary Department
0    Alice  25.0  50000.0         HR
1      Bob   누락1  60000.0    Finance
2  Charlie  35.0      NaN        NaN
3    David  40.0  80000.0         IT
4      Eva   누락1  90000.0  Marketing

In [48]:
df['Salary'] = df['Salary'].fillna(df['Salary'].mean())
df

Name   Age   Salary Department
0    Alice  25.0  50000.0         HR
1      Bob   누락1  60000.0    Finance
2  Charlie  35.0  70000.0        NaN
3    David  40.0  80000.0         IT
4      Eva   누락1  90000.0  Marketing

# 5. 차원 확인

- ndim과 shape 속성 사용
    - ndim은 DataFrame은 2차원 형태를 가지는 경우가 많기  numpy에서 더 자주 사용하게 된다.
- ndim은 차원의 수를 반환하고 shape은 데이터의 형태를 반환


In [45]:
df.shape # 행, 열 확인

(5, 4)

In [46]:
df.ndim # 차원확인

2

# 6. 실습 01

## 01-1
- 데이터를 다운받아서 파일을 읽어오고 다음과 같은 문제에 답하세요.
- nba 데이터의 차원의 수, 모양, 컬럼 인덱스, 로우 인덱스를 각각 구해보세요.


1. 차원의 수 

2. 모양

3. 칼럼 인덱스

4. 로우 인덱스

## 01-2

- nba 데이터에 결측치를 갖는 컬럼이 존재하는지 확인해 봅시다. 
- 그리고 존재한다면 몇개의 데이터나 결측치 값을 갖는지 확인해보세요.
- 또한 각 컬럼별 결측치를 제외한 데이터의 수를 확인해보세요.



## 01-3

- nba 데이터에 각 컬럼마다 고유한 값을 몇개씩 갖는지 조회해보세요. 
- Salary 컬럼의 평균, 최대 최소 등 다양한 통계량을 확인해보세요.


# 7. 고급 인덱싱 

- Pandas는 NumPy 배열과 같이 콤마(,)를 사용한 (row 인덱스, column 인덱스) 형식의 2차원 인덱싱을 지원하기 위해 \
다음과 같은 특별한 인덱서(indexer) 속성도 제공한다.
    - loc : label 값 기반의 2차원 인덱싱
    - iloc : 순서를 나타내는 정수 기반의 2차원 인덱싱


In [49]:
# 10x10 데이터프레임 생성 

data = np.arange(1, 101).reshape(10, 10)
index = [f'row_{i+1}' for i in range(10)]
columns = [f'col_{i+1}' for i in range(10)]

df = pd.DataFrame(data, columns=columns, index=index)

df

col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  col_10
row_1       1      2      3      4      5      6      7      8      9      10
row_2      11     12     13     14     15     16     17     18     19      20
row_3      21     22     23     24     25     26     27     28     29      30
row_4      31     32     33     34     35     36     37     38     39      40
row_5      41     42     43     44     45     46     47     48     49      50
row_6      51     52     53     54     55     56     57     58     59      60
row_7      61     62     63     64     65     66     67     68     69      70
row_8      71     72     73     74     75     76     77     78     79      80
row_9      81     82     83     84     85     86     87     88     89      90
row_10     91     92     93     94     95     96     97     98     99     100

## 7.1 loc

- location의 약자 
- label-based-indeixing
    - 행과 열을 라벨을 사용하여 선택할 때 사용 
- df.loc[row인덱스]
- df.loc[row인덱스, column인덱스]

In [50]:
df.loc['row_1']

col_1      1
col_2      2
col_3      3
col_4      4
col_5      5
col_6      6
col_7      7
col_8      8
col_9      9
col_10    10
Name: row_1, dtype: int64

In [51]:
df.loc['row_1', 'col_1']

np.int64(1)

In [52]:
df.loc[['row_1', 'row_3']]

col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  col_10
row_1      1      2      3      4      5      6      7      8      9      10
row_3     21     22     23     24     25     26     27     28     29      30

In [56]:
df.loc['row_1':'row_3']
df

col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  col_10
row_1       1      2      3      4      5      6      7      8      9      10
row_2      11     12     13     14     15     16     17     18     19      20
row_3      21     22     23     24     25     26     27     28     29      30
row_4      31     32     33     34     35     36     37     38     39      40
row_5      41     42     43     44     45     46     47     48     49      50
row_6      51     52     53     54     55     56     57     58     59      60
row_7      61     62     63     64     65     66     67     68     69      70
row_8      71     72     73     74     75     76     77     78     79      80
row_9      81     82     83     84     85     86     87     88     89      90
row_10     91     92     93     94     95     96     97     98     99     100

In [58]:
df['row_1':'row_3']
# df['컬롬묭'] : 칼럼기반
#df : 로우 기반

col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  col_10
row_1      1      2      3      4      5      6      7      8      9      10
row_2     11     12     13     14     15     16     17     18     19      20
row_3     21     22     23     24     25     26     27     28     29      30

In [61]:
bool_sr = df.col_1 > 50
bool_sr

row_1     False
row_2     False
row_3     False
row_4     False
row_5     False
row_6      True
row_7      True
row_8      True
row_9      True
row_10     True
Name: col_1, dtype: bool

In [62]:
df.loc[bool_sr] # 제외하고 부르기

col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  col_10
row_6      51     52     53     54     55     56     57     58     59      60
row_7      61     62     63     64     65     66     67     68     69      70
row_8      71     72     73     74     75     76     77     78     79      80
row_9      81     82     83     84     85     86     87     88     89      90
row_10     91     92     93     94     95     96     97     98     99     100

In [63]:
df.loc[bool_sr].col_1

row_6     51
row_7     61
row_8     71
row_9     81
row_10    91
Name: col_1, dtype: int64

## 7.2 iloc

- integer-location-based-indexing의 약자 
- 정수 기반 위치 인덱싱
- 행과 열을 정수 기반 위치를 사용해 인덱싱 및 슬라이싱 할 때 사용 


In [65]:
data = np.arange(1,101).reshape(10,10)

index = [f'row_{i}' for i in range(10)]
colums = [f'col{i}' for i in range(10)]

df = pd.DataFrame(data, index=index, columns=columns)
df


col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  col_10
row_0      1      2      3      4      5      6      7      8      9      10
row_1     11     12     13     14     15     16     17     18     19      20
row_2     21     22     23     24     25     26     27     28     29      30
row_3     31     32     33     34     35     36     37     38     39      40
row_4     41     42     43     44     45     46     47     48     49      50
row_5     51     52     53     54     55     56     57     58     59      60
row_6     61     62     63     64     65     66     67     68     69      70
row_7     71     72     73     74     75     76     77     78     79      80
row_8     81     82     83     84     85     86     87     88     89      90
row_9     91     92     93     94     95     96     97     98     99     100

In [66]:
df.iloc[0] # 로우 기반, 반환값은 시지즈

col_1      1
col_2      2
col_3      3
col_4      4
col_5      5
col_6      6
col_7      7
col_8      8
col_9      9
col_10    10
Name: row_0, dtype: int64

In [67]:
df.iloc[1:2]

col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  col_10
row_1     11     12     13     14     15     16     17     18     19      20

In [68]:
df.iloc[1, 3] # 대괄호를 따로 열필요없이 바로 , 로 처리가능

np.int64(14)

In [69]:
# 1~4행, 3~5열 슬라이싱
df.iloc[1:5, 3:6]

col_4  col_5  col_6
row_1     14     15     16
row_2     24     25     26
row_3     34     35     36
row_4     44     45     46

In [70]:
df.iloc[1:4, 2:7:2] # 증감도 가능

col_3  col_5  col_7
row_1     13     15     17
row_2     23     25     27
row_3     33     35     37

## 7.3 조건 필터링 

In [73]:
# col_5 50보다 큰 행만 필터링
df[df['col_5'] > 50]

col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  col_10
row_5     51     52     53     54     55     56     57     58     59      60
row_6     61     62     63     64     65     66     67     68     69      70
row_7     71     72     73     74     75     76     77     78     79      80
row_8     81     82     83     84     85     86     87     88     89      90
row_9     91     92     93     94     95     96     97     98     99     100

In [75]:
# col_5가 30보다 크고, col_7이 70보다 작은 행 필터링

df[(df['col_5'] > 30) & (df['col_7'] < 70)]

col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  col_10
row_3     31     32     33     34     35     36     37     38     39      40
row_4     41     42     43     44     45     46     47     48     49      50
row_5     51     52     53     54     55     56     57     58     59      60
row_6     61     62     63     64     65     66     67     68     69      70

In [76]:
df.loc[(df['col_5'] > 30) & (df['col_7'] < 70)] # 조건 걸시 loc 과 방식이 동일

col_1  col_2  col_3  col_4  col_5  col_6  col_7  col_8  col_9  col_10
row_3     31     32     33     34     35     36     37     38     39      40
row_4     41     42     43     44     45     46     47     48     49      50
row_5     51     52     53     54     55     56     57     58     59      60
row_6     61     62     63     64     65     66     67     68     69      70

In [79]:
# loc과 []만 사용한 경우 추가 작업 가능 여부 확인

# []사용한 경우
display(df[(df['col_5'] > 30) & (df['col_7'] < 70)]['col_1']) # 시리즈로 반환

display(df[(df['col_5'] > 30) & (df['col_7'] < 70), 'col_1']) # 에러 발생

row_3    31
row_4    41
row_5    51
row_6    61
Name: col_1, dtype: int64

InvalidIndexError: (row_0    False
row_1    False
row_2    False
row_3     True
row_4     True
row_5     True
row_6     True
row_7    False
row_8    False
row_9    False
dtype: bool, 'col_1')

In [84]:
# loc 사용한 경우
display(df.loc[(df['col_5'] > 30) & (df['col_7'] < 70)]['col_1']) # 시리즈로 반환

display(df.loc[(df['col_5'] > 30) & (df['col_7'] < 70), ['col_1']]) # 데이터 프레임으로 반환
display(df.loc[(df['col_5'] > 30) & (df['col_7'] < 70), ['col_1', 'col_2']]) # 여러개 가능

display(df.loc[(df['col_5'] > 30) & (df['col_7'] < 70), 'col_1']) # 시리즈로 반환



row_3    31
row_4    41
row_5    51
row_6    61
Name: col_1, dtype: int64

col_1
row_3     31
row_4     41
row_5     51
row_6     61

col_1  col_2
row_3     31     32
row_4     41     42
row_5     51     52
row_6     61     62

row_3    31
row_4    41
row_5    51
row_6    61
Name: col_1, dtype: int64

# 8. 실습 02 

- 타이타닉호 승객의 평균 나이를 구하세요. 29.6
- 타이타닉호 승객중 여성 승객의 평균 나이를 구하세요. 28.7
- 타이타닉호 승객중 1등실(class=="First") 선실의 여성 승객의 평균 나이를 구하세요. 34.3
    - 평균은 df['칼럼명'].mean()을 통해 구할 수 있음 
